tests/federation_room_join_test.go

```diff
@@ -192,3 +194,124 @@ func TestJoinFederatedRoomWithUnverifiableEvents(t *testing.T) {
 		alice.JoinRoom(t, roomAlias, nil)
 	})
 }
+
+// This test checks that we cannot submit anything via /v1/send_join except a join.
+func TestCannotSendNonJoinViaSendJoinV1(t *testing.T) {
+	testValidationForSendMembershipEndpoint(t, "/_matrix/federation/v1/send_join", "join", nil)
+}
+
+// This test checks that we cannot submit anything via /v2/send_join except a join.
+func TestCannotSendNonJoinViaSendJoinV2(t *testing.T) {
+	testValidationForSendMembershipEndpoint(t, "/_matrix/federation/v2/send_join", "join", nil)
+}
+
+// This test checks that we cannot submit anything via /v1/send_leave except a leave.
+func TestCannotSendNonLeaveViaSendLeaveV1(t *testing.T) {
+	testValidationForSendMembershipEndpoint(t, "/_matrix/federation/v1/send_leave", "leave", nil)
+}
+
+// This test checks that we cannot submit anything via /v2/send_leave except a leave.
+func TestCannotSendNonLeaveViaSendLeaveV2(t *testing.T) {
+	testValidationForSendMembershipEndpoint(t, "/_matrix/federation/v2/send_leave", "leave", nil)
+}
+
+// testValidationForSendMembershipEndpoint attempts to submit a range of events via the given endpoint
+// and checks that they are all rejected.
+func testValidationForSendMembershipEndpoint(t *testing.T, baseApiPath, expectedMembership string, createRoomOpts map[string]interface{}) {
+	if createRoomOpts == nil {
+		createRoomOpts = map[string]interface{}{}
```

Comment:
```suggestion
		createRoomOpts = make(map[string]interface{})
```

Not a functional change but this is more idiomatic when you're not shoving content inside of the map when instantiating it.